### PR TITLE
Refactoring MKCOL and extended collections

### PR DESCRIPTION
**Pull request by evert**
The changes we made earlier for PROPPATCH can also be applied to MKCOL.
This patch makes MKCOL and PROPPATCH more consistent with each other,
and also allows the PropertyStorage plugin to hook into MKCOL correctly.

It also renames CardDAV\UserAddressBooks to CardDAV\AddressBookHome to be consistent with CalDAV\CalendarHome.

### CODE REVIEW

**Comment by Hywan** [comment]
Sounds good for me.

File: lib/CalDAV/Plugin.php [hunk 1]
```diff
@@ -287,7 +288,7 @@ function httpMkCalendar(RequestInterface $request, ResponseInterface $response)
             $resourceType = ['{DAV:}collection','{urn:ietf:params:xml:ns:caldav}calendar'];
         }
 
-        $this->server->createCollection($path,$resourceType,$properties);
+        $this->server->createCollection($path,new MkCol($resourceType, $properties));
```
Comment (DominikTo) [inline lib/CalDAV/Plugin.php]: Missing space. ;-)

File: lib/CardDAV/AddressBookHome.php [hunk 2]
```diff
@@ -155,18 +156,20 @@ function getChildren() {
     }
 
     /**
-     * Creates a new addressbook
+     * Creates a new calendar or subscription.
```
Comment (DominikTo) [inline lib/CardDAV/AddressBookHome.php]: Shouldn't it still be `addressbook` in here?

File: lib/DAV/IExtendedCollection.php [hunk 3]
```diff
@@ -15,14 +15,30 @@
 interface IExtendedCollection extends ICollection {
 
     /**
-     * Creates a new collection
+     * Creates a new collection.
+     *
+     * This method will receive MkCol object with all the information about the
```
Comment (DominikTo) [inline lib/DAV/IExtendedCollection.php]: receive `a` MkCol object

File: lib/DAV/IExtendedCollection.php [hunk 4]
```diff
@@ -15,14 +15,30 @@
...
+     * throw Exception\InvalidResourceType.
+     *
+     * The object also contains a list of WebDAV properties for the new
+     * collection.
+     *
+     * You should call the handle() method on this object to specify exactly
+     * which properties you are storing. This allows the system to figure out
+     * exactly which properties you didn't store, which in turn allows other
+     * plugins (such as the propertystorage plugin) to handle storing the
+     * property for you.
```
Comment (DominikTo) [inline lib/DAV/IExtendedCollection.php]: Or maybe:
"properties you didn't handle for you."

File: lib/DAV/Server.php [hunk 5]
```diff
@@ -1095,40 +1095,32 @@ function updateFile($uri,$data, &$etag = null) {
      */
     function createDirectory($uri) {
 
-        $this->createCollection($uri,['{DAV:}collection'], []);
+        $this->createCollection($uri,new MkCol(['{DAV:}collection'], []));
```
Comment (DominikTo) [inline lib/DAV/Server.php]: Missing space. :)

File: lib/DAV/Server.php [hunk 6]
```diff
@@ -1148,72 +1138,51 @@ function createCollection($uri, array $resourceType, array $properties) {
...
         if (!$this->emit('beforeBind',[$uri])) return;
 
-        // There are 2 modes of operation. The standard collection
-        // creates the directory, and then updates properties
-        // the extended collection can create it directly.
         if ($parent instanceof IExtendedCollection) {
 
-            $parent->createExtendedCollection($newName, $resourceType, $properties);
+            /**
+             * If the parent is an instanced of IExtendedCollection, it means that
```
Comment (DominikTo) [inline lib/DAV/Server.php]: instanced -> instance

File: lib/DAV/Server.php [hunk 7]
```diff
@@ -1148,72 +1138,51 @@ function createCollection($uri, array $resourceType, array $properties) {
...
+            $parent->createExtendedCollection($newName, $mkCol);
 
         } else {
 
-            // No special resourcetypes are supported
-            if (count($resourceType)>1) {
+            /**
+             * If the parent is a standard ICollection, it means only
+             * 'standard' collections can be created, so we should fail any
+             * MKCOL operation that carry extra resourcetypes.
```
Comment (DominikTo) [inline lib/DAV/Server.php]: operations that carry
or
operation that carries

**Pull request merged**